src/gsMultiGrid/gsGridHierarchy.hpp

```diff
@@ -0,0 +1,411 @@
+    index_t multiplicityOfKnotsToBeInserted
+    )
+{
+    gsGridHierarchy<T> result;
+    result.m_boundaryConditions = boundaryConditions,
+    result.m_assemblerOptions = assemblerOptions,
+    result.m_mBases.resize(levels);
+    result.m_transferMatrices.resize(levels-1);
+    result.m_localTransferMatrices.resize(levels-1);
+    result.m_mBases[0] = give(mBasis);
+    for ( index_t i=1; i<levels; ++i )
+        uniformRefine_withTransfer(
+            result.m_mBases[i-1],
+            result.m_boundaryConditions,
+            result.m_assemblerOptions,
+            numberOfKnotsToBeInserted,
+            multiplicityOfKnotsToBeInserted,
+            result.m_mBases[i],
+            result.m_transferMatrices[i-1],
+            result.m_localTransferMatrices[i-1]
+        );
+    return result;
+}
+
+template <typename T>
+index_t sizeOfMultiBasis( const gsMultiBasis<T>& mb )
+{
+    const index_t nBases = mb.nBases();
+    index_t result = 0;
+    for (index_t i = 0; i<nBases; ++i)
+        result += mb[i].size();
+    return result;
+}
+
+template <typename T>
+gsGridHierarchy<T> gsGridHierarchy<T>::buildByCoarsening(
+    gsMultiBasis<T> mBasis,
+    const gsBoundaryConditions<T>& boundaryConditions,
+    const gsOptionList& assemblerOptions,
+    index_t levels,
+    index_t degreesOfFreedom
+    )
+{
+    gsGridHierarchy<T> result;
+    result.m_boundaryConditions = boundaryConditions,
+    result.m_assemblerOptions = assemblerOptions,
+
+    result.m_mBases.push_back(give(mBasis));
+
+    index_t lastSize = sizeOfMultiBasis(result.m_mBases[0]);
+
+    for (int i = 0; i < levels && lastSize > degreesOfFreedom; ++i)
+    {
+        gsMultiBasis<T> coarseMBasis;
+        gsSparseMatrix<T, RowMajor> transferMatrix;
+        std::vector< gsSparseMatrix<T, RowMajor> > localTransferMatrices;
+        coarsenMultiBasis_withTransfer(
+            result.m_mBases[i],
+            boundaryConditions,
+            assemblerOptions,
+            coarseMBasis,
+            transferMatrix,
+            localTransferMatrices
+        );
+
+        index_t newSize = sizeOfMultiBasis(coarseMBasis);
+        // If the number of dofs could not be decreased, then cancel. However, if only the number
+        // of levels was specified, then this should be ignored (the caller might need to have a
+        // fixed number of levels).
+        if (lastSize <= newSize && degreesOfFreedom > 0)
+             break;
+        lastSize = newSize;
+
+        result.m_mBases.push_back(give(coarseMBasis));
+        result.m_transferMatrices.push_back(give(transferMatrix));
+        result.m_localTransferMatrices.push_back(give(localTransferMatrices));
+    }
+
+    std::reverse( result.m_mBases.begin(), result.m_mBases.end() );
+    std::reverse( result.m_transferMatrices.begin(), result.m_transferMatrices.end() );
+    std::reverse( result.m_localTransferMatrices.begin(), result.m_localTransferMatrices.end() );
+
+    return result;
+}
+
+/*template <typename T>
```

Comment:
Add a comment on why this function is commented out (eg. future work, debugging, or other)

Reply:
removed; code in devel only for now